### PR TITLE
add example with reactphp

### PR DESCRIPTION
**Pull request by Chrisdowson**

### CODE REVIEW

**Comment by spawnia** [comment]
Static analysis fails: https://github.com/webonyx/graphql-php/runs/7018012660?check_suite_focus=true

**Comment by spawnia** [comment]
Try running static analysis locally, see https://github.com/webonyx/graphql-php/blob/master/CONTRIBUTING.md#static-analysis

File: composer.json [hunk 1]
```diff
@@ -11,7 +11,8 @@
   "require": {
     "php": "^7.4 || ^8",
     "ext-json": "*",
-    "ext-mbstring": "*"
+    "ext-mbstring": "*",
+    "react/http": "^1.6"
```
Comment (spawnia) [inline composer.json:15]: Should be in require-dev

File: examples/04-hello-world-with-reactphp/README.md [hunk 2]
```diff
@@ -0,0 +1,18 @@
+## Hello World Example With Reactphp
+
+This is a Simple yet full-featured example with reactphp.
```
Comment (spawnia) [inline examples/04-hello-world-with-reactphp/README.md:3]:
```suggestion
This is a basic example using [ReactPHP](https://reactphp.org).
```

File: examples/04-hello-world-with-reactphp/graphql.php [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+// curl -d '{"query": "mutation { sum(x: 2, y: 2) }" }' -H "Content-Type: application/json" http://localhost:8080
+
+require_once __DIR__ . '/../../vendor/autoload.php';
+
+use GraphQL\GraphQL;
+use GraphQL\Type\Definition\ObjectType;
+use GraphQL\Type\Definition\Type;
+use GraphQL\Type\Schema;
+
+    $queryType = new ObjectType([
```
Comment (spawnia) [inline examples/04-hello-world-with-reactphp/graphql.php:19]: This and the following code should not be indented

File: examples/04-hello-world-with-reactphp/graphql.php [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+        'name' => 'Query',
+        'fields' => [
+            'echo' => [
+                'type' => Type::string(),
+                'args' => [
+                    'message' => ['type' => Type::string()],
+                ],
+                'resolve' => function ($rootValue, array $args) {
+                    $deferred = new \React\Promise\Deferred();
```
Comment (spawnia) [inline examples/04-hello-world-with-reactphp/graphql.php:28]: use imports

File: examples/04-hello-world-with-reactphp/graphql.php [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+                    $promise = $deferred->promise();
+                    $promise = $promise->then(function () use ($rootValue, $args) {
+                        return $rootValue['prefix'] . $args['message'];
+                    });
```
Comment (spawnia) [inline examples/04-hello-world-with-reactphp/graphql.php:32]:
```suggestion
                    $promise = $promise->then(static fn (): string => $rootValue['prefix'] . $args['message']);
```

File: examples/04-hello-world-with-reactphp/graphql.php [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+
+     // See docs on schema options:
+    // https://webonyx.github.io/graphql-php/schema-definition/#configuration-options
+    $schema = new Schema([
+        'query' => $queryType,
+        'mutation' => $mutationType,
+    ]);
+
+    $react = new \GraphQL\Executor\Promise\Adapter\ReactPromiseAdapter();
+    $server = new \React\Http\HttpServer(function (Psr\Http\Message\ServerRequestInterface $request) use ($schema, $react) {
```
Comment (spawnia) [inline examples/04-hello-world-with-reactphp/graphql.php:63]: Return type?

File: examples/04-hello-world-with-reactphp/graphql.php [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+
+        return $promise->then(function (GraphQL\Executor\ExecutionResult $result) {
+            $output = $result->toArray(1);
+
+            return new \React\Http\Message\Response(
+                200,
+                [
+                    'Content-Type' => 'text/json',
+                ],
+                (json_encode($output) !== false) ? json_encode($output) : ''
```
Comment (spawnia) [inline examples/04-hello-world-with-reactphp/graphql.php:79]: This encodes twice, very inefficient

File: examples/04-hello-world-with-reactphp/graphql.php [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+    $server = new \React\Http\HttpServer(function (Psr\Http\Message\ServerRequestInterface $request) use ($schema, $react) {
+        $rawInput = (string) $request->getBody();
+        $input = json_decode($rawInput, true);
+        $query = $input['query'];
+        $variableValues = $input['variables'] ?? null;
+        $rootValue = ['prefix' => 'You said: '];
+        $promise = GraphQL::promiseToExecute($react, $schema, $query, $rootValue, null, $variableValues);
+
+        return $promise->then(function (GraphQL\Executor\ExecutionResult $result) {
+            $output = $result->toArray(1);
```
Comment (spawnia) [inline examples/04-hello-world-with-reactphp/graphql.php:72]: Use constants over magic values

File: examples/04-hello-world-with-reactphp/graphql.php [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+                [
+                    'Content-Type' => 'text/json',
+                ],
+                (json_encode($output) !== false) ? json_encode($output) : ''
+            );
+        });
+    });
+    $socket = new \React\Socket\SocketServer('127.0.0.1:8010');
+    $server->listen($socket);
+    echo 'Listening on ' . str_replace('tcp:', 'http:', $socket->getAddress() ?? '') . PHP_EOL;
```
Comment (spawnia) [inline examples/04-hello-world-with-reactphp/graphql.php:85]: Why the fallback? When can this be `null`?

**Review by spawnia — CHANGES_REQUESTED** [review]

File: examples/04-hello-world-with-reactphp/graphql.php [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+$server = new HttpServer(function (ServerRequestInterface $request) use ($schema, $react) {
+    $rawInput = (string) $request->getBody();
+    $input = json_decode($rawInput, true);
+    $query = $input['query'];
+    $variableValues = $input['variables'] ?? null;
+    $rootValue = ['prefix' => 'You said: '];
+    $promise = GraphQL::promiseToExecute($react, $schema, $query, $rootValue, null, $variableValues);
+
+    return $promise->then(function (ExecutionResult $result) {
+        $output = json_encode($result->toArray(1));
```
Comment (spawnia) [inline examples/04-hello-world-with-reactphp/graphql.php:73]: Still using a magic value, there is a constant for this

**Comment by spawnia** [comment]
@Chrisdowson I was willing to include this example with a few minor tweaks, if you are still up for it.

**Comment by Chrisdowson** [comment]
> @Chrisdowson I was willing to include this example with a few minor tweaks, if you are still up for it.
of course,It's my pleasure

**Comment by Chrisdowson** [comment]
@spawnia of course, It's my pleasure

**Pull request closed without merging**